cirq-google/cirq_google/ops/gateset.py

```diff
@@ -0,0 +1,112 @@
+from typing import cast, Optional, Union, Type, List, Sequence
+
+import cirq
+from cirq_google.ops import fsim_gate_family, sycamore_gate
+from cirq.protocols.decompose_protocol import DecomposeResult
+import cirq_google.transformers.target_gatesets as cg_target_gatesets
+
+
+def _build_equivalence_gate_families(device_gateset: cirq.Gateset) -> List[cirq.GateFamily]:
+    fsim_gates = [
+        cast(fsim_gate_family.POSSIBLE_FSIM_GATES, g)
+        for g in device_gateset.gates
+        if isinstance(g, tuple(fsim_gate_family.POSSIBLE_FSIM_GATE_TYPES))
+    ]
+    return [fsim_gate_family.FSimGateFamily(gates_to_accept=fsim_gates)]
+
+    # TODO(verult) PhasedXZGateFamily
+
+
+class Gateset(cirq.CompilationTargetGateset):
+    """Represents the gateset of a Google device.
+
+    This gateset accept circuits that use any Cirq gate representation of valid device gates.
+
+    It can also be used as an argument in cirq.optimize_for_target_gateset() to transform your
+    circuit into one that consists of valid gates for the target device. In cases where multiple
+    `CompilationTargetGateset`s are possible for a given gateset, a pre-determined default will be
+    applied, but all available `CompilationTargetGateset`s can be accessed via the `target_gatesets`
+    property.
+
+    Users typically interact with this class by accessing its instantiations via
+    `GoogleDevice.metadata.gateset`.
+
+    Examples:
+
+    ```
+    # View what gates and `CompilationTargetGatesets` are supported by the device.
+    print(gateset)
+
+    # Checks whether a gate is valid for the device.
+    cirq.X in gateset
+
+    # Checks whether the circuit only contains valid gates for the device.
+    gateset.validate(circuit)
+
+    # Transform the circuit into one that can be executed on the device.
+    cirq.optimize_for_target_gateset(circuit, gateset=gateset)
+
+    # Choose your own target gateset, if the device supports multiple.
+    cirq.optimize_for_target_gateset(circuit, gateset=gateset.target_gatesets[1])
+    """
+
+    def __init__(
+        self,
+        device_name,
+        *device_gates: Union[Type[cirq.Gate], cirq.Gate, cirq.GateFamily],
+    ) -> None:
+        self._device_name = device_name
+        self._device_gateset = cirq.Gateset(*device_gates)
+        equivalence_gate_families = _build_equivalence_gate_families(self._device_gateset)
+        super().__init__(*(tuple(self._device_gateset.gates) + tuple(equivalence_gate_families)))
```

Comment:
IIUC, this sets `self._gates` to be the input gates plus an FSimGateFamily over all FSim-type gates. Is that correct?